Content.Client/SubFloor/TrayScannerSystem.cs

```diff
@@ -0,0 +1,141 @@
+using Robust.Shared.IoC;
+using Robust.Shared.Log;
+using Robust.Shared.Maths;
+using Robust.Shared.Utility;
+
+namespace Content.Client.SubFloor;
+
+public class TrayScannerSystem : SharedTrayScannerSystem
+{
+    [Dependency] private IEntityLookup _entityLookup = default!;
+    [Dependency] private SubFloorHideSystem _subfloorSystem = default!;
+    [Dependency] private SharedContainerSystem _containerSystem = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+    }
+
+    public override void ToggleTrayScanner(EntityUid uid, bool toggle, TrayScannerComponent? scanner = null)
+    {
+        if (!Resolve(uid, ref scanner))
+            return;
+
+        scanner.Toggled = toggle;
+        UpdateTrayScanner(uid, scanner);
+
+        if (toggle) _activeScanners.Add(uid);
+    }
+
+    private HashSet<EntityUid> _activeScanners = new();
+    private RemQueue<EntityUid> _invalidScanners = new();
+
+    public override void Update(float frameTime)
+    {
+        if (!_activeScanners.Any()) return;
+
+        foreach (var scanner in _activeScanners)
+        {
+            if (!UpdateTrayScanner(scanner))
+                _invalidScanners.Add(scanner);
+        }
+
+        foreach (var invalidScanner in _invalidScanners)
+            _activeScanners.Remove(invalidScanner);
+
+        if (_invalidScanners.List != null) _invalidScanners.List.Clear();
+    }
+
+    /// <summary>
+    ///     Updates a T-Ray scanner. Should be called on immediate
+    ///     state change (turned on/off), or during the update
+    ///     loop.
+    /// </summary>
+    /// <returns>true if the update was successful, false otherwise</returns>
+    private bool UpdateTrayScanner(EntityUid uid, TrayScannerComponent? scanner = null, TransformComponent? transform = null)
+    {
+        // whoops?
+        if (!Resolve(uid, ref scanner, ref transform))
+        {
+            return false;
+        }
+
+        // if the scanner was toggled off recently,
+        // set all the known subfloor to invisible,
+        // and return false so it's removed from
+        // the active scanner list
+        if (!scanner.Toggled)
+        {
+            _subfloorSystem.ToggleSubfloorEntities(scanner.RevealedSubfloors, false, uid, _visualizerKeys);
+            scanner.LastLocation = Vector2.Zero;
+            scanner.RevealedSubfloors.Clear();
+            return false;
+        }
+
+        // get the rounded position so that small movements don't cause this to
+        // update every time
+        Vector2 flooredPos;
+
+        // zero vector implies container
+        if (transform.LocalPosition == Vector2.Zero
+            && transform.Parent != null
+            && _containerSystem.ContainsEntity(transform.Parent.Owner.Uid, uid))
+        {
+            flooredPos = transform.Parent.LocalPosition.Rounded();
+        }
+        else
+        {
+            flooredPos = transform.LocalPosition.Rounded();
+        }
```

Comment:
There's an issue here where if you put the scanner inside your backpack (a container in a container), then the scanner never thinks it moves away from (0,0) and the revealed entities get frozen in place.

I think the easiest way to fix this would be to just disable the scanner in nested storage. I.e., check if transform.Parent is also in a container, and if it is turn the scanner off. 

Otherwise, I think you just have to get the coordinates relative to a local grid (if there even is one nearby).